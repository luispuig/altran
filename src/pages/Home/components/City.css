._City { margin-top: 65px}

._City ._CardBody {
	min-height: 65px;
	color: #FFFFFF;
	padding-top: 0;
	padding-bottom: 0;
}

._City .frame-temperature {
	text-align: center;
	margin-top: -50px;
	margin-bottom: 10px;
}
._City .temperature {
	background: #FFFFFF;
	width: 100px;
	height: 100px;
	display: inline-block;
	-webkit-border-radius: 50%;
	-moz-border-radius: 50%;
	border-radius: 50%;
	line-height: 100px;
	font-size: 28px;
	font-weight: 200;
}

._City .temperature .unit { font-size: 50% }

._City .error { color: darkred }



