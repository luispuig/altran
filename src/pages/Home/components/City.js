// @flow
import * as React from "react";

import Spinner  from "react-activity/lib/Spinner/";
import "react-activity/lib/Spinner/Spinner.css";

import "./City.css";

type Temperature = {
  id: number,
  date: Date,
  temperature: number
};

type _Props = {
  id: number,
  name: string,
  state: "request" | "success" | "fail",
  temperature?: number,
  error?: string,
  temperatures: Array<_Temperature>
};

class City extends React.Component<_Props> {

  _renderTemperature = () => {
    const {state, temperature} = this.props;
    switch(state) {
      case "request": // Loading
        return <Spinner />;

      case "success":
        if(temperature === undefined) return null;
        return ( <div>{ Math.round(temperature) }
              &nbsp;<span className="unit">ºC</span>
            </div>
        );

      case "fail":
        return <div className="error">X</div> ;

      default:
        return null;
    }
  };

  // Aux function to format date dd/MM/yyyy HH:mm
  // TODO IMPROVE Move function to utils file
  _formatDate = (date:Date) => {
    const dd = ("00" + date.getDate()).slice(-2);
    const mm = ("00" + date.getMonth()).slice(-2);
    const yyyy = date.getFullYear();
    const _hh = ("00" + date.getHours()).slice(-2);
    const _mm = ("00" + date.getMinutes()).slice(-2);

    return `${dd}/${mm}/${yyyy} ${_hh}:${_mm}`;
  };

  render() {
    const { name, state, error, temperatures } = this.props;
    return (
        <div className="_City card elevation-1 elevation-hover">
          <div className="card-body card-img-top bg_color_main _CardBody">
            <div className="frame-temperature">
              <div className="temperature elevation-2 text_color_dark">
                { this._renderTemperature() }
              </div>
            </div>
            <div className="card-title">{name}</div>
          </div>
          <div className="list-group list-group-flush list-temperatures">
            {
              // Check and show error message
              (state === "fail") && error &&
              <li className="list-group-item error">
                {error}
              </li>
            }
            {
              temperatures.map( (item:_Temperature, index:number) =>
                  <li key={index} className="list-group-item">
                    <div className="row">
                      <div className="col-7 col-md-8">
                        {this._formatDate(item.date)}
                      </div>
                      <div className="col-5 col-md-4 text-right">
                        { Math.round(item.temperature) }
                        &nbsp;<span className="unit">ºC</span></div>
                    </div>
                  </li>
              )
            }
          </div>
        </div>
    );
  }
}

export default City;
