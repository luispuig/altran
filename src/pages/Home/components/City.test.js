import React from 'react';
import ReactDOM from 'react-dom';

import ShallowRenderer from 'react-test-renderer/shallow'; // ES6


import City from './City';

const initialProps =
    {
      id: 123,
      name: "CityName",
      state: "success",
      temperature: 20
    }


const  temperatures = [
  {id:3936456,  temperature:15.45,  date: new Date('2017-12-06T10:57:16.462Z')},
  {id:3448439,  temperature:22.06,  date: new Date('2017-12-06T10:57:16.449Z')},
  {id:3871336,  temperature:12,     date:new Date('2017-12-06T10:57:16.438Z')}
];


it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<City {...initialProps} temperatures={temperatures}/>, div);
});

it('match on Shallow mode for REQUEST', () => {
  const props = {...initialProps, state: "request" }
  const renderer = new ShallowRenderer();
  renderer.render(<City {...props} temperatures={temperatures}/>);
  const tree = renderer.getRenderOutput();
  expect(tree).toMatchSnapshot();
});

it('match on Shallow mode for SUCCESS', () => {
  const renderer = new ShallowRenderer();
  renderer.render(<City {...initialProps} temperatures={temperatures}/>);
  const tree = renderer.getRenderOutput();
  expect(tree).toMatchSnapshot();
});

it('match on Shallow mode for SUCCESS. temperatire undefined', () => {
  const renderer = new ShallowRenderer();
  renderer.render(<City {...initialProps } temperature={undefined} temperatures={temperatures}/>);
  const tree = renderer.getRenderOutput();
  expect(tree).toMatchSnapshot();
});

it('match on Shallow mode for ERROR WITHOUT MESSAGE', () => {
  const props = {...initialProps, state: "fail" }
  const renderer = new ShallowRenderer();
  renderer.render(<City {...props} temperatures={temperatures}/>);
  const tree = renderer.getRenderOutput();
  expect(tree).toMatchSnapshot();
});

it('match on Shallow mode for ERROR WITH MESSAGE', () => {
  const props = {...initialProps, state: "fail", error:"Error message" }
  const renderer = new ShallowRenderer();
  renderer.render(<City {...props} temperatures={temperatures}/>);
  const tree = renderer.getRenderOutput();
  expect(tree).toMatchSnapshot();
});

it('match on Shallow mode for NO STATE', () => {
  const props = {...initialProps, state: "NO-STATE", error:"Error message" }
  const renderer = new ShallowRenderer();
  renderer.render(<City {...props} temperatures={temperatures}/>);
  const tree = renderer.getRenderOutput();
  expect(tree).toMatchSnapshot();
});
