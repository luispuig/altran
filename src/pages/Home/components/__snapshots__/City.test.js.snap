// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match on Shallow mode for ERROR WITH MESSAGE 1`] = `
<div
  className="_City card elevation-1 elevation-hover"
>
  <div
    className="card-body card-img-top bg_color_main _CardBody"
  >
    <div
      className="frame-temperature"
    >
      <div
        className="temperature elevation-2 text_color_dark"
      >
        <div
          className="error"
        >
          X
        </div>
      </div>
    </div>
    <div
      className="card-title"
    >
      CityName
    </div>
  </div>
  <div
    className="list-group list-group-flush list-temperatures"
  >
    <li
      className="list-group-item error"
    >
      Error message
    </li>
    <li
      className="list-group-item"
    >
      <div
        className="row"
      >
        <div
          className="col-7 col-md-8"
        >
          06/11/2017 11:57
        </div>
        <div
          className="col-5 col-md-4 text-right"
        >
          15
           
          <span
            className="unit"
          >
            ºC
          </span>
        </div>
      </div>
    </li>
    <li
      className="list-group-item"
    >
      <div
        className="row"
      >
        <div
          className="col-7 col-md-8"
        >
          06/11/2017 11:57
        </div>
        <div
          className="col-5 col-md-4 text-right"
        >
          22
           
          <span
            className="unit"
          >
            ºC
          </span>
        </div>
      </div>
    </li>
    <li
      className="list-group-item"
    >
      <div
        className="row"
      >
        <div
          className="col-7 col-md-8"
        >
          06/11/2017 11:57
        </div>
        <div
          className="col-5 col-md-4 text-right"
        >
          12
           
          <span
            className="unit"
          >
            ºC
          </span>
        </div>
      </div>
    </li>
  </div>
</div>
`;

exports[`match on Shallow mode for ERROR WITHOUT MESSAGE 1`] = `
<div
  className="_City card elevation-1 elevation-hover"
>
  <div
    className="card-body card-img-top bg_color_main _CardBody"
  >
    <div
      className="frame-temperature"
    >
      <div
        className="temperature elevation-2 text_color_dark"
      >
        <div
          className="error"
        >
          X
        </div>
      </div>
    </div>
    <div
      className="card-title"
    >
      CityName
    </div>
  </div>
  <div
    className="list-group list-group-flush list-temperatures"
  >
    <li
      className="list-group-item"
    >
      <div
        className="row"
      >
        <div
          className="col-7 col-md-8"
        >
          06/11/2017 11:57
        </div>
        <div
          className="col-5 col-md-4 text-right"
        >
          15
           
          <span
            className="unit"
          >
            ºC
          </span>
        </div>
      </div>
    </li>
    <li
      className="list-group-item"
    >
      <div
        className="row"
      >
        <div
          className="col-7 col-md-8"
        >
          06/11/2017 11:57
        </div>
        <div
          className="col-5 col-md-4 text-right"
        >
          22
           
          <span
            className="unit"
          >
            ºC
          </span>
        </div>
      </div>
    </li>
    <li
      className="list-group-item"
    >
      <div
        className="row"
      >
        <div
          className="col-7 col-md-8"
        >
          06/11/2017 11:57
        </div>
        <div
          className="col-5 col-md-4 text-right"
        >
          12
           
          <span
            className="unit"
          >
            ºC
          </span>
        </div>
      </div>
    </li>
  </div>
</div>
`;

exports[`match on Shallow mode for NO STATE 1`] = `
<div
  className="_City card elevation-1 elevation-hover"
>
  <div
    className="card-body card-img-top bg_color_main _CardBody"
  >
    <div
      className="frame-temperature"
    >
      <div
        className="temperature elevation-2 text_color_dark"
      />
    </div>
    <div
      className="card-title"
    >
      CityName
    </div>
  </div>
  <div
    className="list-group list-group-flush list-temperatures"
  >
    <li
      className="list-group-item"
    >
      <div
        className="row"
      >
        <div
          className="col-7 col-md-8"
        >
          06/11/2017 11:57
        </div>
        <div
          className="col-5 col-md-4 text-right"
        >
          15
           
          <span
            className="unit"
          >
            ºC
          </span>
        </div>
      </div>
    </li>
    <li
      className="list-group-item"
    >
      <div
        className="row"
      >
        <div
          className="col-7 col-md-8"
        >
          06/11/2017 11:57
        </div>
        <div
          className="col-5 col-md-4 text-right"
        >
          22
           
          <span
            className="unit"
          >
            ºC
          </span>
        </div>
      </div>
    </li>
    <li
      className="list-group-item"
    >
      <div
        className="row"
      >
        <div
          className="col-7 col-md-8"
        >
          06/11/2017 11:57
        </div>
        <div
          className="col-5 col-md-4 text-right"
        >
          12
           
          <span
            className="unit"
          >
            ºC
          </span>
        </div>
      </div>
    </li>
  </div>
</div>
`;

exports[`match on Shallow mode for REQUEST 1`] = `
<div
  className="_City card elevation-1 elevation-hover"
>
  <div
    className="card-body card-img-top bg_color_main _CardBody"
  >
    <div
      className="frame-temperature"
    >
      <div
        className="temperature elevation-2 text_color_dark"
      >
        <Spinner
          animationDuration={0.6}
          speed={1}
        />
      </div>
    </div>
    <div
      className="card-title"
    >
      CityName
    </div>
  </div>
  <div
    className="list-group list-group-flush list-temperatures"
  >
    <li
      className="list-group-item"
    >
      <div
        className="row"
      >
        <div
          className="col-7 col-md-8"
        >
          06/11/2017 11:57
        </div>
        <div
          className="col-5 col-md-4 text-right"
        >
          15
           
          <span
            className="unit"
          >
            ºC
          </span>
        </div>
      </div>
    </li>
    <li
      className="list-group-item"
    >
      <div
        className="row"
      >
        <div
          className="col-7 col-md-8"
        >
          06/11/2017 11:57
        </div>
        <div
          className="col-5 col-md-4 text-right"
        >
          22
           
          <span
            className="unit"
          >
            ºC
          </span>
        </div>
      </div>
    </li>
    <li
      className="list-group-item"
    >
      <div
        className="row"
      >
        <div
          className="col-7 col-md-8"
        >
          06/11/2017 11:57
        </div>
        <div
          className="col-5 col-md-4 text-right"
        >
          12
           
          <span
            className="unit"
          >
            ºC
          </span>
        </div>
      </div>
    </li>
  </div>
</div>
`;

exports[`match on Shallow mode for SUCCESS 1`] = `
<div
  className="_City card elevation-1 elevation-hover"
>
  <div
    className="card-body card-img-top bg_color_main _CardBody"
  >
    <div
      className="frame-temperature"
    >
      <div
        className="temperature elevation-2 text_color_dark"
      >
        <div>
          20
           
          <span
            className="unit"
          >
            ºC
          </span>
        </div>
      </div>
    </div>
    <div
      className="card-title"
    >
      CityName
    </div>
  </div>
  <div
    className="list-group list-group-flush list-temperatures"
  >
    <li
      className="list-group-item"
    >
      <div
        className="row"
      >
        <div
          className="col-7 col-md-8"
        >
          06/11/2017 11:57
        </div>
        <div
          className="col-5 col-md-4 text-right"
        >
          15
           
          <span
            className="unit"
          >
            ºC
          </span>
        </div>
      </div>
    </li>
    <li
      className="list-group-item"
    >
      <div
        className="row"
      >
        <div
          className="col-7 col-md-8"
        >
          06/11/2017 11:57
        </div>
        <div
          className="col-5 col-md-4 text-right"
        >
          22
           
          <span
            className="unit"
          >
            ºC
          </span>
        </div>
      </div>
    </li>
    <li
      className="list-group-item"
    >
      <div
        className="row"
      >
        <div
          className="col-7 col-md-8"
        >
          06/11/2017 11:57
        </div>
        <div
          className="col-5 col-md-4 text-right"
        >
          12
           
          <span
            className="unit"
          >
            ºC
          </span>
        </div>
      </div>
    </li>
  </div>
</div>
`;

exports[`match on Shallow mode for SUCCESS. temperatire undefined 1`] = `
<div
  className="_City card elevation-1 elevation-hover"
>
  <div
    className="card-body card-img-top bg_color_main _CardBody"
  >
    <div
      className="frame-temperature"
    >
      <div
        className="temperature elevation-2 text_color_dark"
      />
    </div>
    <div
      className="card-title"
    >
      CityName
    </div>
  </div>
  <div
    className="list-group list-group-flush list-temperatures"
  >
    <li
      className="list-group-item"
    >
      <div
        className="row"
      >
        <div
          className="col-7 col-md-8"
        >
          06/11/2017 11:57
        </div>
        <div
          className="col-5 col-md-4 text-right"
        >
          15
           
          <span
            className="unit"
          >
            ºC
          </span>
        </div>
      </div>
    </li>
    <li
      className="list-group-item"
    >
      <div
        className="row"
      >
        <div
          className="col-7 col-md-8"
        >
          06/11/2017 11:57
        </div>
        <div
          className="col-5 col-md-4 text-right"
        >
          22
           
          <span
            className="unit"
          >
            ºC
          </span>
        </div>
      </div>
    </li>
    <li
      className="list-group-item"
    >
      <div
        className="row"
      >
        <div
          className="col-7 col-md-8"
        >
          06/11/2017 11:57
        </div>
        <div
          className="col-5 col-md-4 text-right"
        >
          12
           
          <span
            className="unit"
          >
            ºC
          </span>
        </div>
      </div>
    </li>
  </div>
</div>
`;
